README.md
# JSConsoleLogger
![license](https://img.shields.io/github/license/Kyo-suke/JSConsoleLogger.svg) ![license](https://img.shields.io/github/release/Kyo-suke/JSConsoleLogger.svg)

## 概要
Javascriptでロギングを行う為のライブラリです。  

console.log等のconsoleメソッドにフックする事ができ、既存のコードを修正せずにロギングを行う事ができます。  

取得したログをダウンロードする機能もあり、Webアプリケーションのユーザーサポートなどに役立てる事ができます。

## 利用方法
[Releases](https://github.com/Kyo-suke/JSConsoleLogger/releases)からスクリプトをダウンロードし、ご自身のプロジェクトに配置して下さい。  
scriptタグで「js-console-logger.min.js」を読み込み利用します。

```html
<script type="text/javascript" src="js-console-logger.min.js">
```

開発・デバッグ用途に非圧縮版を利用する場合は「js-console-logger.js」をご利用下さい。

```html
<script type="text/javascript" src="js-console-logger.js">
```

## APIリファレンス
### JSConsoleLogger.log(obj1 [, obj2, ..., objN])
引数で渡した内容をログ種別「log」としてロギングします。  
任意のタイミングでロギングしたい場合に利用します。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| obj1 ... objN | any | ロギングするオブジェクト。 |

### JSConsoleLogger.info(obj1 [, obj2, ..., objN])
引数で渡した内容をログ種別「info」としてロギングします。  
任意のタイミングでロギングしたい場合に利用します。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| obj1 ... objN | any | ロギングするオブジェクト。 |

### JSConsoleLogger.debug(obj1 [, obj2, ..., objN])
引数で渡した内容をログ種別「debug」としてロギングします。  
任意のタイミングでロギングしたい場合に利用します。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| obj1 ... objN | any | ロギングするオブジェクト。 |

### JSConsoleLogger.warn(obj1 [, obj2, ..., objN])
引数で渡した内容をログ種別「warn」としてロギングします。  
任意のタイミングでロギングしたい場合に利用します。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| obj1 ... objN | any | ロギングするオブジェクト。 |

### JSConsoleLogger.error(obj1 [, obj2, ..., objN])
引数で渡した内容をログ種別「log」としてロギングします。  
任意のタイミングでロギングしたい場合に利用します。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| obj1 ... objN | any | ロギングするオブジェクト。 |

### JSConsoleLogger.clean()
保持しているログを消去します。

### JSConsoleLogger.getLogData()
現在保持しているログデータ配列の参照を取得します。

### JSConsoleLogger.getLogFileBlob([mimeType] [, withBOM] [, lineFeedCode])
ログファイルのBlobオブジェクトを取得します。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| mimeType | string | ログファイルのMIMEタイプを指定します。 <br> 未指定の場合は"text/plain"となります。 |
| withBOM | boolean | ログファイルにBOMを付けるかを指定します。 <br> true: BOMを付与する。false: BOMを付与しない。 <br> 未指定の場合はfalseとなります。 |
| lineFeedCode | string | ログファイルの改行コード("cr" \| "lf" \| "crlf")を指定します。 <br> 未指定の場合は"crlf"となります。 |

### JSConsoleLogger.save([filename] [, mimeType] [, withBOM] [, lineFeedCode])
ログファイルを生成し、ブラウザからダウンロードを行います。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| filename | string | ログファイルのファイル名を指定します。 <br> 以下の特殊文字を入れる事で特殊なパラメーターに置換されます。 <br> %d : 現在の日付。「YYYYmmDD」形式。 <br> %t : 現在の時間。「HHMMSS」形式。 |
| mimeType | string | ログファイルのMIMEタイプを指定します。 <br> 未指定の場合は"text/plain"となります。 |
| withBOM | boolean | ログファイルにBOMを付けるかを指定します。 <br> true: BOMを付与する。false: BOMを付与しない。 <br> 未指定の場合はfalseとなります。 |
| lineFeedCode | string | ログファイルの改行コード("cr" \| "lf" \| "crlf")を指定します。 <br> 未指定の場合は"crlf"となります。 |

### JSConsoleLogger.setLoggingTarget(logType1 [, logType2 ... , logTypeN])
ロギング対象にするログタイプを設定します。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| logType1 ... logTypeN | any | ロギング対象にするログタイプ。 <br> ("log" \| "info" \| "debug" \| "warn" \| "error") |

### JSConsoleLogger.setLogBufferSize(bufferSize)
保持するログの最大数を設定します。  

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| bufferSize | number | 保持するログの最大数。 |

- ログ最大数のデフォルト値は1000です。
- ログの最大数を超えた場合、古いログから削除されます。

### JSConsoleLogger.setStackTraceOnConsole(enabled)
コンソールメソッドをフックすると、開発者がブラウザコンソールから発生箇所が分かりにくくなるデメリットがあります。  
フックした各種コンソールメソッドが実行された際にスタックトレースを表示する事で、この問題を解決する事が出来ます。

| 引数 | 型 | 説明 |
|:---:|:---:|:---|
| enabled | boolean | true : スタックトレースを有効にする。 <br> false : スタックトーレスを無効にする。 |

- デフォルトは無効に設定されています。
- InternetExplorer及びMicrosoft Edgeはこのオプションに対応していません。

## 対応ブラウザ
ECMAScript5以降に対応したモダンブラウザでご利用いただけます。

- Internet Explorerは10以降に対応。

## ライセンス
MIT License。
[LICENSE](LICENSE)ファイルを参照して下さい。

## 更新履歴
詳しい更新履歴は[CHANGELOG.md](CHANGELOG.md)を参照して下さい。


CHANGELOG.md
# CHANGELOG
## v1.0.2
### Changed
- Add timezone date to log.

### Fixed
- Fix to log formatter write wrong month.

## v1.0.1
### Fixed
- Fix problem to fall down into recursive call loop when using console.trace on IE and MSEdge.

## v1.0.0
Initial release.


src/ts/entry.ts
/// <reference path="./common.ts" />
/// <reference path="./logger.ts" />

namespace JSConsoleLogger {
    /**
     * Logging message as a type "log".
     * @param args Logging messages.
     */
    export function log(...args: any[]): void {
        Logger.push("log", args);
    }

    /**
     * Logging message as a type "info".
     * @param args Logging messages.
     */
    export function info(...args: any[]): void {
        Logger.push("info", args);
    }

    /**
     * Logging message as a type "debug".
     * @param args Logging messages.
     */
    export function debug(...args: any[]): void {
        Logger.push("debug", args);
    }

    /**
     * Logging message as a type "warn".
     * @param args Logging messages.
     */
    export function warn(...args: any[]): void {
        Logger.push("warn", args);
    }

    /**
     * Logging message as a type "error".
     * @param args Logging messages.
     */
    export function error(...args: any[]): void {
        Logger.push("error", args);
    }

    /**
     * Clear Log buffer.
     */
    export function clear(): void {
        Logger.clear();
    }

    /**
     * Get log data list.
     * @return log data object list.
     */
    export function getLogData(): LogData[] {
        return Logger.getLogData();
    }

    /**
     * Get Log file blob object.
     * @param [mimeType="text/plain"] Specify mime type string.
     * @param [withBOM=false] true : Insert BOM to top of the file.
     * @param [lineFeedCodeType="crlf"] Specify linefeed code type. "cr" | "lf" | "crlf".
     */
    export function getLogFileBlob(mimeType: string = "text/plain", withBOM: boolean = false, lineFeedCodeType: LineFeedCodeType = "crlf"): Blob {
        let logData = Logger.getLogData();
        let logFile = new LogFile(logData, mimeType, withBOM, lineFeedCodeType);
        return logFile.entity;
    }

    /**
     * Save log file to local.
     * @param [filename] Download filename.
     * @param [mimeType="text/plain"] Specify mime type string.
     * @param [withBOM=false] true : Insert BOM to top of the file.
     * @param [lineFeedCodeType="crlf"] Specify linefeed code type. "cr" | "lf" | "crlf".
     */
    export function save(
        filename: string = null,
        mimeType: string = "text/plain",
        withBOM: boolean = false,
        lineFeedCodeType: LineFeedCodeType = "crlf"
    ): void {
        let logData = Logger.getLogData();
        let logFile = new LogFile(logData, mimeType, withBOM, lineFeedCodeType);
        Downloader.download(logFile, filename);
    }

    /**
     * Set filter logging target.
     * @param loggingTargets - Specify Logging target strings.
     */
    export function setLoggingTarget(...loggingTargets: LogType[]): void {
        Logger.setLoggingTarget.apply(Logger, loggingTargets);
    }

    /**
     * Set Logging buffer size.
     * @param bufferSize Number of logs to keep.
     */
    export function setLogBufferSize(bufferSize: number): void {
        Logger.setLogBufferSize(bufferSize);
    }

    /**
     * Hooking a console method can obcure the caller on browser console.
     * The problem is solved by displaying the stack trace
     * when the hooked console method is executed.
     * This feature is off by default.
     * Note : This option is not support Internet Explorer and Microsoft Edge.
     * @param enabled true: show stack trace on hooked console method.
     */
    export function setStackTraceOnConsole(enabled: boolean): void {
        Logger.setStackTraceOnConsole(enabled);
    }
}


src/ts/logger.ts
/// <reference path="./common.ts" />

namespace JSConsoleLogger {
    /**
     * @static
     * @class Logger
     * @classdesc Logging several messages.
     */
    export class Logger {
        /**
         * Max size of log message list.
         */
        private static _bufferSize: number = 1000;
        /**
         * log message list entity.
         */
        private static _buffer: LogData[];
        /**
         * Bits of logging target.
         */
        private static _loggingTarget: LoggingTarget;
        /**
         * Show stack trace on console logging.
         */
        private static _isStackTrace: boolean = false;

        constructor() {
            throw new Error("Logger is static class.");
        }

        /**
         * Initialize method.
         * @static
         * @param [bufferSize=1000] Max size of log message buffer list
         * @param loggingTargets Specify Logging target strings.
         */
        public static initialize(bufferSize: number = 1000, ...loggingTargets: LogType[]): void {
            if (loggingTargets.length <= 0) {
                loggingTargets = ["debug", "error", "info", "log", "warn"];
            }
            this._bufferSize = bufferSize;
            this.clear();
            this.setLoggingTarget.apply(this, loggingTargets);
            this._hook();
        }

        /**
         * Make to empty buffer list.
         * @static
         */
        public static clear(): void {
            this._buffer = [];
        }

        /**
         * Set filter logging target.
         * @static
         * @param loggingTargets Specify Logging target strings.
         */
        public static setLoggingTarget(...loggingTargets: LogType[]): void {
            if (!loggingTargets) {
                return;
            }
            // set logging target bit flag
            this._loggingTarget = 0;
            for (let target of loggingTargets) {
                if (!(target in LoggingTarget)) {
                    continue;
                }
                this._loggingTarget |= LoggingTarget[target];
            }
        }

        /**
         * Set Logging buffer size.
         * @param bufferSize Number of logs to keep.
         */
        public static setLogBufferSize(bufferSize: number): void {
            bufferSize = bufferSize || 0;
            this._bufferSize = bufferSize;
            this._rotate();
        }

        /**
         * Set enable of stack trace on hooked console method.
         * @param enabled true: show stack trace on hooked console method.
         */
        public static setStackTraceOnConsole(enabled: boolean): void {
            this._isStackTrace = enabled;
        }

        /**
         * Append log message to buffer list.
         * @static
         * @param type Log type string
         * @param messages Target of message
         */
        public static push(type: LogType, messages: any[]): void {
            if (!this._isLoggingTarget(type)) {
                return;
            }

            let message = "";
            let delimiter = ", ";
            // formatting arguments taken to console method to string
            for (let i = 0; i < messages.length; i++) {
                let item = messages[i];
                if (i > 0) {
                    message += delimiter;
                }
                switch (typeof item) {
                    // string
                    case typeof "":
                        message += item;
                        break;
                    // number
                    case typeof 0:
                        var _item: number = <number>item;
                        message += _item.toString();
                        break;
                    // object
                    case typeof {}:
                        try {
                            message += JSON.stringify(item);
                        } catch (e) {
                            message += "[object]";
                        }
                    // undefined
                    case typeof undefined:
                        message += "undefined";
                        break;
                    // etc
                    default:
                        message += "";
                        break;
                }
            }
            // append log data
            var logData: LogData = {
                type: type,
                message: message,
                time: new Date()
            };
            this._buffer.push(logData);

            // rotate
            this._rotate();
        }

        /**
         * Log rotate.
         */
        private static _rotate(): void {
            if (this._bufferSize > 0 && this._buffer.length > this._bufferSize) {
                let overSize = this._buffer.length - this._bufferSize;
                this._buffer.splice(0, overSize);
            }
        }

        /**
         * Hook various console method.
         */
        private static _hook(): void {
            let cls = this;
            function _genTraceFunc(type: LogType, origFunc: Function) {
                return function _trace(...args: any[]): void {
                    // fall down into recursive call loop when using console.trace on IE and MSEdge
                    if (_trace.caller) {
                        if (_trace.caller === _trace) {
                            return;
                        }
                    }
                    // original trace
                    origFunc.apply(console, args);
                    // stack trace
                    if (console.trace && cls._isStackTrace) {
                        console.trace();
                    }
                    // push log data
                    cls.push(type, args);
                    return;
                };
            }

            // console is not defined on not developer mode in old Internet Explorer.
            if (!console) {
                return;
            }

            // console.debug
            console.debug = _genTraceFunc("debug", console.debug);
            // console.error
            console.error = _genTraceFunc("error", console.error);
            // console.info
            console.info = _genTraceFunc("info", console.info);
            // console.log
            console.log = _genTraceFunc("log", console.log);
            // console.warn
            console.warn = _genTraceFunc("warn", console.warn);
        }

        /**
         * Check log type contain in logging targets.
         * @param logType Log type string
         * @return Return true When log type contain in logging target
         */
        private static _isLoggingTarget(logType: LogType): boolean {
            if (!(logType in LoggingTarget)) {
                return false;
            }
            let bits = LoggingTarget[logType];
            return (this._loggingTarget & bits) > 0;
        }

        public static getLogData(): LogData[] {
            return this._buffer;
        }
    }

    Logger.initialize();
}


src/ts/logfile.ts
/// <reference path="./common.ts" />
/// <reference path="./logformatter.ts" />

namespace JSConsoleLogger {
    /**
     * @class LogFile
     * @classdesc Log file data object.
     */
    export class LogFile {
        private static readonly BOM: Uint8Array = new Uint8Array([0xef, 0xbb, 0xbf]);

        private _mimeType: string;
        private _lineFeedCodeType: LineFeedCodeType;
        private _withBOM: boolean;
        private _entity: Blob;

        /**
         * @param datas Target Log data object list.
         * @param [mimeType="text/plain"] Specify mime type string.
         * @param [withBOM=false] true : Insert BOM to top of the file.
         * @param [lineFeedCodeType="crlf"] Specify linefeed code type. "cr" | "lf" | "crlf".
         */
        constructor(datas: LogData[], mimeType: string = "text/plain", withBOM: boolean = false, lineFeedCodeType: LineFeedCodeType = "crlf") {
            let cls = LogFile;
            this._mimeType = mimeType;
            this._lineFeedCodeType = lineFeedCodeType;
            this._withBOM = withBOM;

            var contents: any[] = this._logData2StrList(datas);
            if (this._withBOM) {
                contents.unshift(cls.BOM);
            }

            this._entity = new Blob(contents, { type: this._mimeType });
        }

        /**
         * Log data convert to string list.
         * @param datas log data list
         * @return converted string list
         */
        private _logData2StrList(datas: LogData[]): string[] {
            let ret = [];
            for (let i = 0; i < datas.length; i++) {
                let logString = LogFormatter.format(datas[i]);
                // add linefeed code
                logString += "\n";
                logString = this._replaceLineFeedCode(logString);
                ret.push(logString);
            }
            return ret;
        }

        /**
         * Unify linefeed code.
         * @param target target string
         */
        private _replaceLineFeedCode(target: string): string {
            let ret = "";
            switch (this._lineFeedCodeType) {
                case "lf":
                    ret = target.replace("\r\n", "\n");
                    break;
                case "cr":
                    ret = target.replace("\r\n", "\r");
                    break;
                case "crlf":
                    ret = target.replace("\r\n", "\n");
                    ret = ret.replace("\r", "\n");
                    ret = ret.replace("\n", "\r\n");
                    break;
                default:
                    return target;
            }
            return ret;
        }

        get entity(): Blob {
            return this._entity;
        }
    }
}


dest/js-console-logger.js
/**
 * JSConsoleLogger v1.0.2
 * Released under the MIT license
 * https://github.com/Kyo-suke/JSConsoleLogger
 */
var JSConsoleLogger;
(function (JSConsoleLogger) {
    var LoggingTarget;
    (function (LoggingTarget) {
        LoggingTarget[LoggingTarget["debug"] = 1] = "debug";
        LoggingTarget[LoggingTarget["error"] = 2] = "error";
        LoggingTarget[LoggingTarget["info"] = 4] = "info";
        LoggingTarget[LoggingTarget["log"] = 8] = "log";
        LoggingTarget[LoggingTarget["warn"] = 16] = "warn";
    })(LoggingTarget = JSConsoleLogger.LoggingTarget || (JSConsoleLogger.LoggingTarget = {}));
})(JSConsoleLogger || (JSConsoleLogger = {}));
var JSConsoleLogger;
(function (JSConsoleLogger) {
    var Downloader = (function () {
        function Downloader() {
            throw new Error("Downloader is static class.");
        }
        Downloader.download = function (logFile, filenameBase) {
            var _this = this;
            var blob = logFile.entity;
            if (!blob) {
                return;
            }
            var filename = this._genFilename(filenameBase);
            if (navigator.msSaveBlob) {
                navigator.msSaveBlob(blob, filename);
                return;
            }
            var url = null;
            try {
                var element = this._createLinkElement();
                url = URL.createObjectURL(blob);
                this._setAttributes(element, filename, url);
                element.click();
                setTimeout(function () {
                    _this._revoke(url);
                }, this.REVOKE_TIMEOUT);
            }
            catch (e) {
                console.error("failed to download.");
                console.error(e.meesage);
                console.error(e.stack);
                this._removeLinkElement();
                if (url) {
                    this._revoke(url);
                }
            }
        };
        Downloader._revoke = function (url) {
            URL.revokeObjectURL(url);
            this._removeLinkElement();
        };
        Downloader._genFilename = function (filenameBase) {
            if (!filenameBase) {
                filenameBase = this.DEFAULT_FILENAME;
            }
            var filename = filenameBase;
            var date = new Date();
            var d = [date.getFullYear(), date.getMonth(), date.getDate()].join("");
            filename = filename.replace("%d", d);
            var t = [date.getHours(), date.getMinutes(), date.getSeconds()].join("");
            filename = filename.replace("%t", t);
            return filename;
        };
        Downloader._createLinkElement = function () {
            var element = document.querySelector("#" + this.ELEMENT_ID);
            if (element) {
                this._removeLinkElement();
            }
            var a = document.createElement("a");
            a.setAttribute("id", this.ELEMENT_ID);
            a.setAttribute("download", "");
            a.setAttribute("target", "_blank");
            a.setAttribute("href", "");
            a.setAttribute("style", "display: block; visibility: hidden; width: 0; height: 0;");
            return document.body.appendChild(a);
        };
        Downloader._removeLinkElement = function () {
            var element = document.querySelector("#" + this.ELEMENT_ID);
            if (element) {
                document.body.removeChild(element);
            }
        };
        Downloader._setAttributes = function (element, filename, href) {
            element.setAttribute("download", filename);
            element.setAttribute("href", href);
            return element;
        };
        Downloader.ELEMENT_ID = "JSConsoleLogDownloader";
        Downloader.DEFAULT_FILENAME = "log_%d_%t.log";
        Downloader.REVOKE_TIMEOUT = 5000;
        return Downloader;
    }());
    JSConsoleLogger.Downloader = Downloader;
})(JSConsoleLogger || (JSConsoleLogger = {}));
var JSConsoleLogger;
(function (JSConsoleLogger) {
    var Logger = (function () {
        function Logger() {
            throw new Error("Logger is static class.");
        }
        Logger.initialize = function (bufferSize) {
            if (bufferSize === void 0) { bufferSize = 1000; }
            var loggingTargets = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                loggingTargets[_i - 1] = arguments[_i];
            }
            if (loggingTargets.length <= 0) {
                loggingTargets = ["debug", "error", "info", "log", "warn"];
            }
            this._bufferSize = bufferSize;
            this.clear();
            this.setLoggingTarget.apply(this, loggingTargets);
            this._hook();
        };
        Logger.clear = function () {
            this._buffer = [];
        };
        Logger.setLoggingTarget = function () {
            var loggingTargets = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                loggingTargets[_i] = arguments[_i];
            }
            if (!loggingTargets) {
                return;
            }
            this._loggingTarget = 0;
            for (var _a = 0, loggingTargets_1 = loggingTargets; _a < loggingTargets_1.length; _a++) {
                var target = loggingTargets_1[_a];
                if (!(target in JSConsoleLogger.LoggingTarget)) {
                    continue;
                }
                this._loggingTarget |= JSConsoleLogger.LoggingTarget[target];
            }
        };
        Logger.setLogBufferSize = function (bufferSize) {
            bufferSize = bufferSize || 0;
            this._bufferSize = bufferSize;
            this._rotate();
        };
        Logger.setStackTraceOnConsole = function (enabled) {
            this._isStackTrace = enabled;
        };
        Logger.push = function (type, messages) {
            if (!this._isLoggingTarget(type)) {
                return;
            }
            var message = "";
            var delimiter = ", ";
            for (var i = 0; i < messages.length; i++) {
                var item = messages[i];
                if (i > 0) {
                    message += delimiter;
                }
                switch (typeof item) {
                    case typeof "":
                        message += item;
                        break;
                    case typeof 0:
                        var _item = item;
                        message += _item.toString();
                        break;
                    case typeof {}:
                        try {
                            message += JSON.stringify(item);
                        }
                        catch (e) {
                            message += "[object]";
                        }
                    case typeof undefined:
                        message += "undefined";
                        break;
                    default:
                        message += "";
                        break;
                }
            }
            var logData = {
                type: type,
                message: message,
                time: new Date()
            };
            this._buffer.push(logData);
            this._rotate();
        };
        Logger._rotate = function () {
            if (this._bufferSize > 0 && this._buffer.length > this._bufferSize) {
                var overSize = this._buffer.length - this._bufferSize;
                this._buffer.splice(0, overSize);
            }
        };
        Logger._hook = function () {
            var cls = this;
            function _genTraceFunc(type, origFunc) {
                return function _trace() {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i] = arguments[_i];
                    }
                    if (_trace.caller) {
                        if (_trace.caller === _trace) {
                            return;
                        }
                    }
                    origFunc.apply(console, args);
                    if (console.trace && cls._isStackTrace) {
                        console.trace();
                    }
                    cls.push(type, args);
                    return;
                };
            }
            if (!console) {
                return;
            }
            console.debug = _genTraceFunc("debug", console.debug);
            console.error = _genTraceFunc("error", console.error);
            console.info = _genTraceFunc("info", console.info);
            console.log = _genTraceFunc("log", console.log);
            console.warn = _genTraceFunc("warn", console.warn);
        };
        Logger._isLoggingTarget = function (logType) {
            if (!(logType in JSConsoleLogger.LoggingTarget)) {
                return false;
            }
            var bits = JSConsoleLogger.LoggingTarget[logType];
            return (this._loggingTarget & bits) > 0;
        };
        Logger.getLogData = function () {
            return this._buffer;
        };
        Logger._bufferSize = 1000;
        Logger._isStackTrace = false;
        return Logger;
    }());
    JSConsoleLogger.Logger = Logger;
    Logger.initialize();
})(JSConsoleLogger || (JSConsoleLogger = {}));
var JSConsoleLogger;
(function (JSConsoleLogger) {
    function log() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        JSConsoleLogger.Logger.push("log", args);
    }
    JSConsoleLogger.log = log;
    function info() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        JSConsoleLogger.Logger.push("info", args);
    }
    JSConsoleLogger.info = info;
    function debug() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        JSConsoleLogger.Logger.push("debug", args);
    }
    JSConsoleLogger.debug = debug;
    function warn() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        JSConsoleLogger.Logger.push("warn", args);
    }
    JSConsoleLogger.warn = warn;
    function error() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        JSConsoleLogger.Logger.push("error", args);
    }
    JSConsoleLogger.error = error;
    function clear() {
        JSConsoleLogger.Logger.clear();
    }
    JSConsoleLogger.clear = clear;
    function getLogData() {
        return JSConsoleLogger.Logger.getLogData();
    }
    JSConsoleLogger.getLogData = getLogData;
    function getLogFileBlob(mimeType, withBOM, lineFeedCodeType) {
        if (mimeType === void 0) { mimeType = "text/plain"; }
        if (withBOM === void 0) { withBOM = false; }
        if (lineFeedCodeType === void 0) { lineFeedCodeType = "crlf"; }
        var logData = JSConsoleLogger.Logger.getLogData();
        var logFile = new JSConsoleLogger.LogFile(logData, mimeType, withBOM, lineFeedCodeType);
        return logFile.entity;
    }
    JSConsoleLogger.getLogFileBlob = getLogFileBlob;
    function save(filename, mimeType, withBOM, lineFeedCodeType) {
        if (filename === void 0) { filename = null; }
        if (mimeType === void 0) { mimeType = "text/plain"; }
        if (withBOM === void 0) { withBOM = false; }
        if (lineFeedCodeType === void 0) { lineFeedCodeType = "crlf"; }
        var logData = JSConsoleLogger.Logger.getLogData();
        var logFile = new JSConsoleLogger.LogFile(logData, mimeType, withBOM, lineFeedCodeType);
        JSConsoleLogger.Downloader.download(logFile, filename);
    }
    JSConsoleLogger.save = save;
    function setLoggingTarget() {
        var loggingTargets = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            loggingTargets[_i] = arguments[_i];
        }
        JSConsoleLogger.Logger.setLoggingTarget.apply(JSConsoleLogger.Logger, loggingTargets);
    }
    JSConsoleLogger.setLoggingTarget = setLoggingTarget;
    function setLogBufferSize(bufferSize) {
        JSConsoleLogger.Logger.setLogBufferSize(bufferSize);
    }
    JSConsoleLogger.setLogBufferSize = setLogBufferSize;
    function setStackTraceOnConsole(enabled) {
        JSConsoleLogger.Logger.setStackTraceOnConsole(enabled);
    }
    JSConsoleLogger.setStackTraceOnConsole = setStackTraceOnConsole;
})(JSConsoleLogger || (JSConsoleLogger = {}));
var JSConsoleLogger;
(function (JSConsoleLogger) {
    var LogFormatter = (function () {
        function LogFormatter() {
            throw new Error("LogFormatter is static class.");
        }
        LogFormatter.format = function (logData) {
            var logTypeStr = this._getLogTypeStr(logData.type);
            var dateStr = this._getDataStr(logData.time);
            var timeStr = this._getTimeStr(logData.time);
            var messageStr = logData.message || "";
            return [dateStr, timeStr, logTypeStr, messageStr].join(this.DELIMITER);
        };
        LogFormatter._getLogTypeStr = function (logType) {
            return "[" + logType + "]";
        };
        LogFormatter._getDataStr = function (targetDate) {
            var ret = "";
            try {
                var y = this._paddingStr(targetDate.getFullYear(), "0", 4);
                var m = this._paddingStr(targetDate.getMonth() + 1, "0", 2);
                var d = this._paddingStr(targetDate.getDate(), "0", 2);
                ret = [y, m, d].join("/");
            }
            catch (e) {
                ret = "--/--/--";
            }
            return ret;
        };
        LogFormatter._getTimeStr = function (targetDate) {
            var ret = "";
            try {
                var h = this._paddingStr(targetDate.getHours(), "0", 2);
                var m = this._paddingStr(targetDate.getMinutes(), "0", 2);
                var s = this._paddingStr(targetDate.getSeconds(), "0", 2);
                var tz = this._getTimezoneStr(targetDate);
                ret = [h, m, s].join(":");
                ret += " " + tz;
            }
            catch (e) {
                ret = "??:??:?? +0000";
            }
            return ret;
        };
        LogFormatter._getTimezoneStr = function (d) {
            var num = Math.abs((d.getTimezoneOffset() * 100) / 60);
            var polarity = d.getTimezoneOffset() > 0 ? "-" : "+";
            return polarity + this._paddingStr(num, "0", 4);
        };
        LogFormatter._paddingStr = function (target, paddingChar, length) {
            var ret = "";
            try {
                var paddingStr = new Array(length + 1).join(paddingChar.slice(0, 1));
                ret = (paddingStr + target).slice(-1 * length);
            }
            catch (e) {
                ret = "" + target;
            }
            return ret;
        };
        LogFormatter.DELIMITER = " ";
        return LogFormatter;
    }());
    JSConsoleLogger.LogFormatter = LogFormatter;
})(JSConsoleLogger || (JSConsoleLogger = {}));
var JSConsoleLogger;
(function (JSConsoleLogger) {
    var LogFile = (function () {
        function LogFile(datas, mimeType, withBOM, lineFeedCodeType) {
            if (mimeType === void 0) { mimeType = "text/plain"; }
            if (withBOM === void 0) { withBOM = false; }
            if (lineFeedCodeType === void 0) { lineFeedCodeType = "crlf"; }
            var cls = LogFile;
            this._mimeType = mimeType;
            this._lineFeedCodeType = lineFeedCodeType;
            this._withBOM = withBOM;
            var contents = this._logData2StrList(datas);
            if (this._withBOM) {
                contents.unshift(cls.BOM);
            }
            this._entity = new Blob(contents, { type: this._mimeType });
        }
        LogFile.prototype._logData2StrList = function (datas) {
            var ret = [];
            for (var i = 0; i < datas.length; i++) {
                var logString = JSConsoleLogger.LogFormatter.format(datas[i]);
                logString += "\n";
                logString = this._replaceLineFeedCode(logString);
                ret.push(logString);
            }
            return ret;
        };
        LogFile.prototype._replaceLineFeedCode = function (target) {
            var ret = "";
            switch (this._lineFeedCodeType) {
                case "lf":
                    ret = target.replace("\r\n", "\n");
                    break;
                case "cr":
                    ret = target.replace("\r\n", "\r");
                    break;
                case "crlf":
                    ret = target.replace("\r\n", "\n");
                    ret = ret.replace("\r", "\n");
                    ret = ret.replace("\n", "\r\n");
                    break;
                default:
                    return target;
            }
            return ret;
        };
        Object.defineProperty(LogFile.prototype, "entity", {
            get: function () {
                return this._entity;
            },
            enumerable: true,
            configurable: true
        });
        LogFile.BOM = new Uint8Array([0xef, 0xbb, 0xbf]);
        return LogFile;
    }());
    JSConsoleLogger.LogFile = LogFile;
})(JSConsoleLogger || (JSConsoleLogger = {}));
